

<!doctype html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>CHANGELOG &#8212; MoP3  documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/bizstyle.css" />
    
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/bizstyle.js"></script>
    <link rel="canonical" href="https://biocorecrg.github.io/MoP3/changelog.html" />
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Benchmark" href="benchmark.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <!--[if lt IE 9]>
    <script src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="benchmark.html" title="Benchmark"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">MoP3  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">CHANGELOG</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="changelog">
<span id="home-page-changelog"></span><h1>CHANGELOG<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h1>
<section id="version-3-0">
<h2>Version 3.0<a class="headerlink" href="#version-3-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><dl class="simple">
<dt>mop_preprocess</dt><dd><ul>
<li><p>We added a custom model for m6A basecalling. It is automatically installed when running INSTALL.sh. For using it you need to indicated <a href="#id1"><span class="problematic" id="id2">``</span></a>–pars_tools “drna_tool_splice_m6A_opt.tsv” ``</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
<section id="version-2-0">
<h2>Version 2.0<a class="headerlink" href="#version-2-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Completely rewritten using the powerful <a class="reference external" href="https://www.nextflow.io/docs/latest/dsl2.html">DSL2</a>.</p></li>
<li><p>Subworkflows are stored in the independent repository <a class="reference external" href="https://github.com/biocorecrg/BioNextflow">BioNextflow</a>.</p></li>
<li><p>Global nextflow config is broken down to different profiles (cluster, cloud, local…)</p></li>
<li><p>Added the new module <strong>mop_consensus</strong></p></li>
<li><dl class="simple">
<dt>mop_preprocess (formerly known as nanoPreprocess + nanoPreprocessSimple)</dt><dd><ul>
<li><p>now can read multiple runs per time using the syntax <strong>“PATH/**/*.fast5”</strong></p></li>
<li><p>can demultiplex fast5 using guppy too</p></li>
<li><p>deeplexicon can be run on GPU too</p></li>
<li><p>Parameters of each tool are stored in a tsv file. We have different ones already pre-set for cDNA, DNA and dRNA (option <strong>–pars_tools</strong>)</p></li>
<li><p>Added new process <strong>discovery</strong> with bambu / isoquant for discovering and quantifying new transcripts.</p></li>
<li><p>demultiplexing, filtering, mapping, counting and discovery can be switched off by setting “NO” as a parameter</p></li>
<li><p>saveSpace can be set to “YES” to reduce the amount of disk space required. <strong>WARNING This will prevent the possibility to resume!</strong></p></li>
<li><p>Merged old NanoPreprocess and NanoPreprocessSimple in <strong>mop_preprocess</strong>. Using fastq or fast5 will switch among the two executions.</p></li>
<li><p>Htseq-count now accepts alignments generated by minimap2. <a class="reference external" href="https://github.com/htseq/htseq/issues/33">https://github.com/htseq/htseq/issues/33</a></p></li>
<li><p>We can specify a <strong>final_summary_</strong>.txt** for extracting kit and flowcell info in the params.config file. If not present we should specify those info or a custom model via extra parameters in one of the <strong>*_opt.tsv</strong> files or guppy will trigger an error.</p></li>
<li><p>This module can be run in AWS BATCH using the profile <strong>awsbatch</strong></p></li>
<li><p>demultiplexing of fast5 with deeplexicon is now faster thanks to multithreading and parallelization</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mop_tail (formerly known as nanoTail)</dt><dd><ul>
<li><p>now you can launch each analysis independently</p></li>
<li><p>Fine tuning of parameter for each step in tools_opt.tsv</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mop_mod (formerly known as nanoMod)</dt><dd><ul>
<li><p>coming SOON!</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</section>
<section id="version-1-1">
<h2>Version 1.1<a class="headerlink" href="#version-1-1" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Added a new module called NanoPreprocessSimple that starts from fastq files instead of fast5 files. It allows the analysis of multiple files at a time.</p></li>
<li><p>Added support to vbz compressed fast5 <a class="reference external" href="https://github.com/nanoporetech/vbz_compression">https://github.com/nanoporetech/vbz_compression</a> in NanoPreprocess, NanoMod and NanoTail</p></li>
<li><p>NanoPreprocess now outputs also CRAM files and can do downsampling with the parameter –downsampling</p></li>
<li><p>NanoPreprocess allows performing variant calling using medaka (BETA)</p></li>
<li><p>NanoPreprocess allows performing demultiplexing with GUPPY</p></li>
<li><p>Added plots for Epinano output in NanoMod</p></li>
<li><p>Added a conversion of Tombo results in bed format in NanoMod</p></li>
<li><p>Added a INSTALL.sh file for automatically retrieve guppy 3.4.5 from <a class="reference external" href="https://mirror.oxfordnanoportal.com/">https://mirror.oxfordnanoportal.com/</a>, place it in NanoPreprocess/bin and making the required links</p></li>
<li><p>Added profiles for being used locally and on the CRG SGE cluster</p></li>
</ul>
</section>
<section id="version-1-0">
<h2>Version 1.0<a class="headerlink" href="#version-1-0" title="Permalink to this headline">¶</a></h2>
<p>This is the original version published in the paper <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fgene.2020.00211/full">MasterOfPores: A Workflow for the Analysis of Oxford Nanopore Direct RNA Sequencing Datasets</a></p>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">CHANGELOG</a><ul>
<li><a class="reference internal" href="#version-3-0">Version 3.0</a></li>
<li><a class="reference internal" href="#version-2-0">Version 2.0</a></li>
<li><a class="reference internal" href="#version-1-1">Version 1.1</a></li>
<li><a class="reference internal" href="#version-1-0">Version 1.0</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="benchmark.html"
                        title="previous chapter">Benchmark</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/changelog.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="benchmark.html" title="Benchmark"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">MoP3  documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">CHANGELOG</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright .
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.3.2.
    </div>
  </body>
</html>