<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>CHANGELOG &mdash; Master of Pores 2  documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Master of Pores 2
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="about.html">About Master of Pores</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">How to install</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_preprocess.html">MOP_PREPROCESS</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_mod.html">MOP_MOD</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_consensus.html">MOP_CONSENSUS</a></li>
<li class="toctree-l1"><a class="reference internal" href="mop_tail.html">MOP_TAIL</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Master of Pores 2</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>CHANGELOG</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="changelog">
<span id="home-page-changelog"></span><h1>CHANGELOG<a class="headerlink" href="#changelog" title="Permalink to this headline"></a></h1>
<section id="version-2-0">
<h2>Version 2.0<a class="headerlink" href="#version-2-0" title="Permalink to this headline"></a></h2>
<p>Completely rewritten using the powerful <a class="reference external" href="https://www.nextflow.io/docs/latest/dsl2.html">DSL2</a>.
Subworkflows are stored in the independent repository <a class="reference external" href="https://github.com/biocorecrg/BioNextflow">BioNextflow</a>.</p>
<ul>
<li><p>mop_preprocess (formerly known as nanoPreprocess + nanoPreprocessSimple)
* now can read multiple runs per time
* can demultiplex fast5 using guppy too
* deeplexicon can be run on GPU too
* Parameters of each tool are stored in a tsv file. We have three different ones already pre-set for cDNA, DNA and dRNA (option –pars_tools)
* demultiplexing, filtering, mapping and counting can be switched off by setting “NO” as a parameter
* saveSpace can be set to “YES” to reduce the amount of disk space required. WARNING This will prevent the possibility to resume!
* Merged old NanoPreprocess and NanoPreprocessSimple in a mop_preprocess. Using fastq or fast5 will switch among the two executions.
* Added new process “discovery” with bambu / … for discovering and quantifying new transcripts.</p></li>
<li><p>mop_mod (formerly known as nanoMod)
* now you can launch each analysis independently
* 4 workflows based on the following tools:</p>
<blockquote>
<div><ul class="simple">
<li><p>epinano</p></li>
<li><p>nanopolish + nanocompore</p></li>
<li><p>tombo model_sample_compare</p></li>
<li><p>tombo level_sample_compare</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p>Fine tuning of parameter for each step in tools_opt.tsv</p></li>
</ul>
</li>
<li><p>mop_tail (formerly known as nanoTail)
* now you can launch each analysis independently
* 4 workflows: epinano, nanopolish + nanocompore, tombo model_sample_compare and tombo level_sample_compare
* Fine tuning of parameter for each step in tools_opt.tsv</p></li>
<li><p>mop_consensus (new module!!)
This module will use the results from mop_mod using the tools from the 4 workflows (epinano, nanopolish, nanocompore and tombo model_sample_compare) to generate a consensus between them . It analyze each reference molecule independently and in parallel for each comparison.</p></li>
</ul>
</section>
<section id="version-1-1">
<h2>Version 1.1<a class="headerlink" href="#version-1-1" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Added a new module called NanoPreprocessSimple that starts from fastq files instead of fast5 files. It allows the analysis of multiple files at a time.</p></li>
<li><p>Added support to vbz compressed fast5 <a class="reference external" href="https://github.com/nanoporetech/vbz_compression">https://github.com/nanoporetech/vbz_compression</a> in NanoPreprocess, NanoMod and NanoTail</p></li>
<li><p>NanoPreprocess now outputs also CRAM files and can do downsampling with the parameter –downsampling</p></li>
<li><p>NanoPreprocess allows performing variant calling using medaka (BETA)</p></li>
<li><p>NanoPreprocess allows performing demultiplexing with GUPPY</p></li>
<li><p>Added plots for Epinano output in NanoMod</p></li>
<li><p>Added a conversion of Tombo results in bed format in NanoMod</p></li>
<li><p>Added a INSTALL.sh file for automatically retrieve guppy 3.4.5 from <a class="reference external" href="https://mirror.oxfordnanoportal.com/">https://mirror.oxfordnanoportal.com/</a>, place it in NanoPreprocess/bin and making the required links</p></li>
<li><p>Added profiles for being used locally and on the CRG SGE cluster</p></li>
</ul>
</section>
<section id="version-1-0">
<h2>Version 1.0<a class="headerlink" href="#version-1-0" title="Permalink to this headline"></a></h2>
<p>This is the original version published in the paper <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fgene.2020.00211/full">MasterOfPores: A Workflow for the Analysis of Oxford Nanopore Direct RNA Sequencing Datasets</a></p>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, CRG.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>